sys/fido2/ctap/ctap_crypto.c

```diff
@@ -0,0 +1,309 @@
+                                  const uint8_t *in, size_t in_size,
+                                  uint8_t *auth_data, size_t auth_data_size,
+                                  uint8_t mac_size, uint8_t length_encoding,
+                                  const uint8_t *nonce, size_t nonce_size,
+                                  const uint8_t *key, size_t key_size)
+{
+    cipher_t cipher;
+    int ret;
+
+    ret = cipher_init(&cipher, CIPHER_AES_128, key, key_size);
+
+    if (ret != 1) {
+        return CTAP1_ERR_OTHER;
+    }
+
+    ret = cipher_encrypt_ccm(&cipher, auth_data, auth_data_size, mac_size,
+                             length_encoding, nonce, nonce_size,
+                             in, in_size, out);
+
+    if (ret < 0) {
+        return CTAP1_ERR_OTHER;
+    }
+
+    return CTAP2_OK;
+}
+
+int fido2_ctap_crypto_aes_ccm_dec(uint8_t *out, size_t out_size,
+                                  const uint8_t *in, size_t in_size,
+                                  uint8_t *auth_data, size_t auth_data_size,
+                                  uint8_t mac_size, uint8_t length_encoding,
+                                  const uint8_t *nonce, size_t nonce_size,
+                                  const uint8_t *key, size_t key_size)
+{
+    cipher_t cipher;
+    int ret, len;
+
+    ret = cipher_init(&cipher, CIPHER_AES, key, CCM_BLOCK_SIZE);
+
+    if (ret != 1) {
+        return CTAP1_ERR_OTHER;
+    }
+
+    len = cipher_decrypt_ccm(&cipher, auth_data, auth_data_size,
+                             mac_size, length_encoding, nonce, nonce_size,
+                             in, in_size, out);
+
+    if (len < 0) {
+        return CTAP1_ERR_OTHER;
+    }
+
+    return CTAP2_OK;
+}
+
+int fido2_ctap_crypto_gen_keypair(ctap_crypto_pub_key_t *pub_key,
+                                  uint8_t *priv_key)
+{
+    int ret;
+    const struct uECC_Curve_t *curve = uECC_secp256r1();
+
+    ret = uECC_make_key((uint8_t *)pub_key, priv_key, curve);
+    if (ret == 0) {
+        return CTAP1_ERR_OTHER;
+    }
+
+    return CTAP2_OK;
+}
+
+int fido2_ctap_crypto_get_sig(uint8_t *hash, size_t hash_size, uint8_t *sig,
+                              size_t *sig_size, const uint8_t *key,
+                              size_t key_size)
+{
+    assert(*sig_size >= CTAP_CRYPTO_ES256_DER_MAX_SIZE);
+
+    uint8_t r_raw[CTAP_CRYPTO_KEY_SIZE];
+    uint8_t s_raw[CTAP_CRYPTO_KEY_SIZE];
+    int ret;
+
+    const struct uECC_Curve_t *curve = uECC_secp256r1();
+
+    ret = uECC_sign(key, hash, hash_size, sig, curve);
+
+    if (ret == 0) {
+        return CTAP1_ERR_OTHER;
+    }
+
+    memmove(r_raw, sig, sizeof(r_raw));
+    memmove(s_raw, sig + sizeof(r_raw), sizeof(s_raw));
+
+    ret = sig_to_der_format(r_raw, s_raw, sig, sig_size);
+
+    if (ret != CTAP2_OK) {
+        return ret;
+    }
+
+    return CTAP2_OK;
+}
+
+static int RNG(uint8_t *dest, size_t size)
+{
+    fido2_ctap_crypto_prng(dest, size);
+    return 1;
```

Comment:
The idea was to add status codes to ´fido2_ctap_crypto_prng´. Once done, these should be processed here I guess